### PR TITLE
move Floor, Gcd, Lcm, Pi to datafusion-functions

### LINKED ISSUE #9861: move Floor, Gcd, Lcm, Pi to datafusion-functions
### Is your feature request related to a problem or challenge?

As part of https://github.com/apache/arrow-datafusion/issues/9285 move the  Floor, Gcd, Lcm, Pi functions to the new datafusion-functions crate in the new structure.

### Describe the solution you'd like

Functions are migrated to the new crate and use the new patterns as described in https://github.com/apache/arrow-datafusion/issues/9286 & https://github.com/apache/arrow-datafusion/pull/9216

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #9861 

## Rationale for this change

Function migration.

## What changes are included in this PR?

code, tests.

## Are these changes tested?

Yes.

## Are there any user-facing changes?

No.

### CODE REVIEW

**Review by jackwener — COMMENTED** [review]
I think this PR is `behavior change`, so I add label.

As part of https://github.com/apache/arrow-datafusion/issues/9285 (for link EPIC issue)

**Review by jackwener — APPROVED** [review]
Thanks @Omega359

**Comment by Omega359** [comment]
Thanks @jackwener - may I ask what you see as a behaviour change? From my understanding there shouldn't be any user facing changes beyond where a function is located in the code (for dataframe use cases). Did I miss/change something inadvertently (it's happened before ...) ?

File: datafusion/functions/src/math/lcm.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+use datafusion_common::{exec_err, DataFusionError, Result};
+use datafusion_expr::ColumnarValue;
+use datafusion_expr::{ScalarUDFImpl, Signature, Volatility};
+
+use crate::math::gcd::compute_gcd;
+use crate::utils::make_scalar_function;
+
+#[derive(Debug)]
+pub struct LcmFunc {
```
Comment (jackwener) [inline datafusion/functions/src/math/lcm.rs:33]: Should it be`pub(super)` ?

File: datafusion/physical-expr/src/utils/mod.rs [hunk 2]
```diff
@@ -265,10 +267,103 @@ mod tests {
...
-    use datafusion_common::{Result, ScalarValue};
+    use datafusion_common::{exec_err, DataFusionError, Result, ScalarValue};
 
+    use datafusion_expr::{
+        ColumnarValue, FuncMonotonicity, ScalarUDFImpl, Signature, Volatility,
+    };
     use petgraph::visit::Bfs;
 
+    #[derive(Debug, Clone)]
+    pub struct TestScalarUDF {
```
Comment (jackwener) [inline datafusion/physical-expr/src/utils/mod.rs:278]: Should it bepub(super) ?

**Comment by jackwener** [comment]
> Thanks @jackwener - may I ask what you see as a behaviour change? From my understanding there shouldn't be any user facing changes beyond where a function is located in the code (for dataframe use cases). Did I miss/change something inadvertently (it's happened before ...) ?

my mistake, sorry for it.

File: datafusion/functions/src/math/lcm.rs (hunk 1 above)
Comment (Omega359) [reply datafusion/functions/src/math/lcm.rs:33]: I don't believe so - this is following the pattern set out in #9903

**Review by jackwener — APPROVED** [review]
I prepare to merge this PR next day if others don't have any question.

**Review by comphead — APPROVED** [review]
lgtm thanks @Omega359

**Comment by comphead** [comment]
Will merge it before the conflicts

**Pull request merged**